service_level_controller: update configuration on raft change

Jadw1 opened the pull request:
This patch is a follow-up to https://github.com/scylladb/scylladb/pull/16585.

Once we have service levels on raft, we can get rid of update loop, which updates the configuration in a configured interval (default is 10s).
Instead, this PR introduces methods to `group0_state_machine` which look through table ids in mutations in `write_mutation` and update submodules based on that ids.

Fixes: https://github.com/scylladb/scylladb/issues/18060

Jadw1 commented:
Please take a closer look at https://github.com/scylladb/scylladb/pull/18758/commits/dc131033065dd5d64c021aaf5d5ce584f7a22c30:

> service/qos/service_level_controller: add a method to reload configuration
> 
> Instead of updating service level configuration in a loop,
> the method schedules a one time reload of the configuration.
> In case of reload failure, it's scheduled once again.
> 
> TODO:
> - Should there been any sleep between failed reload and th retry?
> - If sth is wrong, should the metod try to do the reload until a success
>   or the number of retries should be limited?

Jadw1 commented on service/qos/service_level_controller.cc:
```diff
@@ -212,6 +217,27 @@ future<> service_level_controller::update_service_levels_from_distributed_data()
+        return;
+    }
+
+    _global_controller_db->configuration_reload = _global_controller_db->configuration_reload.then([this] () -> future<> {
```
I'm wrong here.
`group0_state_machine` should do `co_await update_service_levels_from_distributed_data()` to make sure doing read barrier is equal to updating service levels configuration.

piodul commented on main.cc:
```diff
@@ -1858,6 +1852,8 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             startlog.info("SSTable data integrity checker is {}.",
                     cfg->enable_sstable_data_integrity_check() ? "enabled" : "disabled");
 
+            sl_controller.local().maybe_start_update_loop(feature_service.local(), cfg);
```
You are postponing the first reload of service levels pretty late, after the messaging service is initialized. Won't this cause problems with us not knowing the service levels of the incoming connections and using the default scheduling group for them?

Jadw1 replied:
No, it shouldn't. Comment above removed call says `This starts the update loop - but no real update happens until the data accessor is not initialized.` and data accessor is initialized just after this line.

kostja replied:
Perhaps the name is confusing now and should be changed?

kostja replied:
maybe_start_update_loop().
It is confusing to call this "update loop" after this patch. It's not a loop.

kostja replied:
But now the main case is raft. Maybe rename it to maybe_start_legacey_update_loop()?

piodul commented on main.cc:
```diff
@@ -1834,7 +1829,6 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             auto stop_group0_usage_in_storage_service = defer_verbose_shutdown("group 0 usage in local storage", [&ss] {
                 ss.local().wait_for_group0_stop().get();
             });
-
```
Nit: unnecessary whitespace change

piodul replied:
Not resolved :/

kostja commented on gms/feature_service.hh:
```diff
@@ -138,6 +138,7 @@ public:
     gms::feature group0_schema_versioning { *this, "GROUP0_SCHEMA_VERSIONING"sv };
     gms::feature supports_consistent_topology_changes { *this, "SUPPORTS_CONSISTENT_TOPOLOGY_CHANGES"sv };
     gms::feature host_id_based_hinted_handoff { *this, "HOST_ID_BASED_HINTED_HANDOFF"sv };
+    gms::feature service_levels_raft_notifications { *this, "SERVICE_LEVELS_RAFT_NOTIFICATIONS"sv };
```
Do you really need a separate feature here or can just wire it up to the overall topology-on-raft bit?

piodul replied:
> Do you really need a separate feature here or can just wire it up to the overall topology-on-raft bit?

If this will be included 6.0 then yes, we could reuse the existing feature, but this PR does not assume that this will happen.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9143/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9143/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels::*
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9143/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1715/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9740/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9143/)
#### Build Details:
- Duration: 3 hr 0 min
- Builder: releng-cloud-x86 (i4i.12xlarge)

Jadw1 commented:
Changelog:
- resolved merge conflict

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9296/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9296/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9296/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1795/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9819/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9296/)
#### Build Details:
- Duration: 2 hr 43 min
- Builder: spider1.cloudius-systems.com

piodul commented on service/qos/raft_service_level_distributed_data_accessor.cc:
```diff
@@ -47,12 +52,23 @@ future<qos::service_levels_info> raft_service_level_distributed_data_accessor::g
 
-    auto group0_cmd = _group0_client.prepare_command(change, guard, description);
+future<> raft_service_level_distributed_data_accessor::do_raft_command(service::group0_guard guard, abort_source& as, std::vector<mutation> mutations, std::string_view description) const {    
+    auto change = create_raft_change(_qp.proxy().features(), {mutations.begin(), mutations.end()});
```
> `{mutations.begin(), mutations.end()}`

Huh, I'm surprised that it works. I suppose it invokes a constructor of `std::vector<canonical_mutation>` which takes a pair of iterators.

It's a bit too surprising for my taste. When I see an initializer list used to create a collection, I expect that all the comma-separated items will be separate elements of that collection.

Please change it to something more explicit, e.g. `std::vector<collection_mutation>(mutations.begin(), mutations.end())` or `boost::copy_range<std::vector<collection_mutation>>(mutations.begin(), mutations.end())`.

piodul commented on service/raft/group0_state_machine.cc:
```diff
@@ -169,7 +169,8 @@ future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merg
     [&] (service_levels_change& chng) -> future<> {
-        return write_mutations_to_database(_sp, cmd.creator_addr, std::move(chng.mutations));
+        co_await write_mutations_to_database(_sp, cmd.creator_addr, std::move(chng.mutations));
+        co_await _ss.update_service_levels_configuration();
```
I think a different order of the commits would make more sense: first introduce `update_service_levels_configuration`, then introduce `service_levels_change`. This way you will be able to write a correct handler for `service_levels_change` at once.

piodul commented on service/qos/raft_service_level_distributed_data_accessor.cc:
```diff
@@ -10,10 +10,15 @@
 #include "db/consistency_level_type.hh"
 #include <seastar/core/abort_source.hh>
+#include "seastar/util/variant_utils.hh"
+#include "service/raft/group0_state_machine.hh"
```
>  service/qos/raft_sl_dda: use service_levels_change when enabled
> 
> When RAFT_SERVICE_LEVELS_CHANGE is enabled, use new `service_levels_change` change.
> Otherwise, use old `write_mutations`.

Nit: this description lacks explanations. Why do we want to use the new `service_levels_change`?
(I know why, it's just that it should be explained in the description)

piodul commented on service/storage_service.hh:
```diff
@@ -860,6 +860,8 @@ public:
     // Public for `reload_raft_topology_state` REST API.
     future<> topology_transition();
 
+    future<> update_service_levels_configuration();
```
This function can only be called on shard 0, otherwise an assertion will be triggered.

This is an important precondition which should be stated as a part of the interface. Please put a comment above this declaration mentioning the precondition.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -221,6 +222,24 @@ future<> service_level_controller::update_service_levels_from_distributed_data()
+}
+
+void service_level_controller::stop_update_from_distributed_data() {
+    if (!_global_controller_db || _global_controller_db->dist_data_update_aborter.abort_requested()) {
```
Do I understand correctly that `_global_controller_db != nullptr` if and only if `this_shard_id() == 0`?

If so, then it's not useful to be able to call this function on other shards than shard 0. I think it's better to be intentional and just not let call it on other shards. Let's change it so that it asserts that it's called on shard 0 and mention in the docstring as a precondition.

Jadw1 commented:
Changelog:
- addressed commens (mostly comments, some small style/code improvements, no logic change)
- rebased on latest masters

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9408/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9408/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9408/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9878/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1854/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9408/)
#### Failed Tests (1/32135):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9408/testReport/junit/boost.memtable_test.failed_flush_prevents_writes.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 43 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9409/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9409/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9409/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1855/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9879/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9409/)
#### Build Details:
- Duration: 2 hr 46 min
- Builder: spider6.cloudius-systems.com

Jadw1 commented:
Changelog:
- the update loop now is started when: the feature is not enabled or cluster is in recovery mode or migrate to raft topology hasn't been done yet
- update loop shutdowns itself. on every iteration it checks the conditions and terminates when all of them are met
- if the update loop isn't started, do the initial update of in-memory configuration to fetch existing service levels

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9490/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9490/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9490/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1901/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9926/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9490/)
#### Build Details:
- Duration: 2 hr 50 min
- Builder: spider8.cloudius-systems.com

Jadw1 commented:
Changelog:
Rebased on latest master with merged https://github.com/scylladb/scylladb/pull/17910, which introduced `group0_batch`. Dropped a new raft command type `service_level_change` and related feature.
Instead, added a methods to look through table ids in mutations in `write_mutations` and updates submodules based on that ids.

Jadw1 commented:
@piodul ping for review

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9594/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9594/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9594/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2001/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10025/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9594/)
#### Build Details:
- Duration: 2 hr 37 min
- Builder: spider8.cloudius-systems.com

piodul commented on service/raft/group0_state_machine.cc:
```diff
@@ -140,6 +140,28 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
+future<> group0_state_machine::reload_modules(modules_to_reload modules) {
+    for (auto& module: modules) {
+        switch (module) {
+            case reload_module::service_level_controller:
+                co_await _ss.update_service_levels_configuration();
+                break;
```
Nit: indented too much; we usually don't indent `case` with respect to `switch`

piodul commented on service/raft/group0_state_machine.hh:
```diff
@@ -104,6 +109,8 @@ class group0_state_machine : public raft_state_machine {
     gms::feature::listener_registration _topology_on_raft_support_listener;
 
+    modules_to_reload find_modules_to_reload(const std::vector<canonical_mutation>& mutations);
+    future<> reload_modules(modules_to_reload modules);
```
> raft/group0_state_machine: update submodules based on table_id

@kbr-scylla @gleb-cloudius please review this commit

piodul commented on service/qos/service_level_controller.hh:
```diff
@@ -150,9 +150,10 @@ public:
      * Start legacy update loop if RAFT_SERVICE_LEVELS_CHANGE feature is not enable yet 
      * or the cluster is in recovery mode
      * or the cluster hasn't been migrated to raft topology.
+     * Otherwise, do initial update of service level configuration to fetch already existing data.
```
Why is it necessary to do this? Won't we load the service levels configurations when we start group0?

Jadw1 replied:
No, we don't. But I've moved this to topology_state_load()

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -307,18 +328,24 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
                 try {
                     f.get();
+                    
+                    if (!group0_client.in_recovery() && storage_service.get_topology_upgrade_state() == service::topology::upgrade_state_type::done) {
```
I don't think it's necessary to check for recovery mode here. You are already checking it before starting the loop. Being in recovery or not is something that is determined on startup and doesn't change over the runtime of the node.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -307,18 +328,24 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
                     f.get();
+                    
+                    if (!group0_client.in_recovery() && storage_service.get_topology_upgrade_state() == service::topology::upgrade_state_type::done) {
+                        stop_update_from_distributed_data();
+                        return make_ready_future<stop_iteration>(stop_iteration::yes);
```
I think that we should print some message here. The message should mention that the polling loop is being stopped, but also explain that service level updates will now be immediate and the loop will no longer be necessary.

piodul commented on service/qos/service_level_controller.hh:
```diff
@@ -143,6 +146,14 @@ public:
     void abort_group0_operations();
 
+    /**
+     * Start legacy update loop if RAFT_SERVICE_LEVELS_CHANGE feature is not enable yet 
```
Nit: `not enable` -> `not enabled`

piodul commented on service/raft/group0_state_machine.hh:
```diff
@@ -94,6 +94,11 @@ struct group0_command {
 // Raft state machine implementation for managing group 0 changes (e.g. schema changes).
 // NOTE: group 0 raft server is always instantiated on shard 0.
 class group0_state_machine : public raft_state_machine {
+    enum class reload_module {
+        service_level_controller
+    };
+    using modules_to_reload = std::unordered_set<reload_module>;
```
Nit: check out the `enum_set` class, I think it would be better to use it here

piodul commented on service/raft/group0_state_machine.cc:
```diff
@@ -140,6 +140,28 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
+}
+
+future<> group0_state_machine::reload_modules(modules_to_reload modules) {
+    for (auto& module: modules) {
+        switch (module) {
+            case reload_module::service_level_controller:
```
I imagine that the function should be organized as a sequence of ifs, checking for whether it's necessary to reload some module, and reload it if the appropriate flag is set in the `modules` variable. This will allow us to control the order of reloading things which will be useful later.

piodul reviewed (changes requested):
Some more comments from me, but otherwise LGTM. I'd like for Kamil and Gleb to review the parts that are relevant for them, too.

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -174,7 +196,9 @@ future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merg
         co_return;
     },
     [&] (write_mutations& muts) -> future<> {
```
Here is the specification for `write_mutations` written at the point of definition:
```cpp
// This command is used to write data to tables other than topology or
// schema tables and it doesn't update any in-memory data structures.
struct write_mutations {
    std::vector<canonical_mutation> mutations;
};
```

this new code contradicts that specification. (**and it doesn't update any in-memory data structures**)

kbr-scylla replied:
I think I recall specifically requesting in some discussion, slack or some design doc thread, not to touch `write_mutations` (get off my lawn :P) but create a separate command type

Jadw1 replied:
Previously this PR added a `service_level_change`, but recently @nuivall's PR added `group0_batch` which collects mutations across whole request (e.g. some commands do changes to both auth and service levels and we want to make this change atomically) and post it as `write_mutation`, so I've dropped the new command type.

Another thing is in https://github.com/scylladb/scylladb/pull/19085 (follow up to this PR) I need to update service levels cache on changes to some auth tables but on only to 2 of them and auth has more. 
So I did a commit similar to this and @gleb-cloudius (https://github.com/scylladb/scylladb/pull/19085#discussion_r1629440154) wrote he thought of sth similar, that's why I did it also here.

Some time ago @piodul proposed to introduce
```
struct {
	enum component_type : uint16_t {
		schema,
		topology,
		// In the near future:
		service_levels,
		auth,
	};

	std::vector<component_type> components_to_update;
	std::vector<canonical_mutation> mutations;
};
```
(https://scylladb-corporate.slack.com/archives/CK7BG6Y6L/p1716557376467699) but AFAIR there was no conclusion or there were some arguments against it.

Would it be okay to create a new command type like `notification_change` and do the same thing as this commit is doing with `write_mutations` (browse table_ids of mutations and update defined subsystems)? 
This way `group0_batch` can return a single raft command and I can still "subscribe" to appropriate tables.

kbr-scylla replied:
> Would it be okay to create a new command type like notification_change and do the same thing as this commit is doing with write_mutations (browse table_ids of mutations and update defined subsystems)?

Yes

IIUC this is something that we agreed on to do eventually -- but the scope of that is larger, we want all components to be update-able correctly that way, but that requires refactoring the components so we can apply RCU style updates to them and update all of them at once as a single isolated transaction.

I think @nuivall agreed to push this effort (in exchange for merging his PR :P)

For now I think we can just create the command type and update only service levels, and later extend the logic to be able to update all components. We just need to make sure that we don't corner ourselves by implementing something that won't be extendable later.

@piodul @nuivall did someone take notes / write a design doc based on the discussions we had and decisions we made?

Let's make sure that introducing `notification_change` is compatible with this so `notification_change` can be later extended with whatever @nuivall wants to do

nuivall replied:
yes, I am trying to push it, making some notes for myself, later maybe transfer to the design doc depending on the need.

from my pov do whatever works here and is inline with current code, likely I will need to refactor it anyway, don't see how it can be forward compatible with something which doesn't yet exists.

kbr-scylla replied:
> from my pov do whatever works here and is inline with current code, likely I will need to refactor it anyway, 

It would be nice if we were able to avoid adding yet another command type later when you have your final design.

On a glance the `notification_change` proposed here by @Jadw1 is similar to what you want to do, so I hope it will be forward-compatible.

Alternatively if this PR is not a priority (and I don't think it is?) we can simply delay merging it until the design is ready.

piodul replied:
@kbr-scylla this is an important follow-up for service levels on raft, other work depend on it (https://github.com/scylladb/scylladb/pull/19085). It would delay our current effort of fixing bugs in SLs.

kbr-scylla replied:
Well anyway I'm fine with introducing `notification_change` as described by @Jadw1 . @gleb-cloudius what's your opinion?

kbr-scylla replied:
It's a change that causes notifications, so I think "notification change" (short of: a change of the notification kind) is the appropriate one

Alternatively, `notifying_change`

kostja replied:
Sorry for bikeshedding, but I think the suggested name is confusing. A notification presumes there is a subscriber. The in-memory state of group0 state machine is not a subscriber, it's the object that is being changed. The name "service level change" was good enough.

gleb-cloudius replied:
Let me join the bikeshedding. I do not like either name. They sound like we want to add yet another command type while we want to replace all command types by this new one.  I am also not sure why do we need a new command type instead of extending `write_mutations` like the patch does. This is not external API. We can review the current usage and see if it is safe to do so.

piodul replied:
> Let me join the bikeshedding. I do not like either name. They sound like we want to add yet another command type while we want to replace all command types by this new one. I am also not sure why do we need a new command type instead of extending `write_mutations` like the patch does. This is not external API. We can review the current usage and see if it is safe to do so.

@kbr-scylla's objection to just extending `write_mutations` was that it would violate the current contract (i.e. it doesn't update any in-memory structures). I'm personally fine with changing the contract to allow the current change as we are ultimately going to get rid of `write_mutations` anyway. @kbr-scylla WDYT?

kbr-scylla replied:
**If** you verify that the current usages of `write_mutations` won't be affected.

For example the usage in sending CDC generation data. I don't want those commands to cause any reload of topology state or anything. They should be pure mutation writes.
(It's fine to execute the logic to calculate what needs to be reloaded -- if the result is that nothing needs to be reloaded)

> I'm personally fine with changing the contract to allow the current change as we are ultimately going to get rid of write_mutations anyway. @kbr-scylla WDYT?

Get rid of it, how? You just proposed to extend it so basically going into the direction of making `write_mutations` "the" command responsible for everything in the future.

piodul replied:
> **If** you verify that the current usages of `write_mutations` won't be affected.
> 
> For example the usage in sending CDC generation data. I don't want those commands to cause any reload of topology state or anything. They should be pure mutation writes. (It's fine to execute the logic to calculate what needs to be reloaded -- if the result is that nothing needs to be reloaded)

In that case I believe they won't affect current use because we only reload service levels cache when there are mutations that modify the service levels table. @Jadw1 please confirm.

> Get rid of it, how? You just proposed to extend it so basically going into the direction of making `write_mutations` "the" command responsible for everything in the future.

Ok, I didn't phrase what I wanted to say well. I meant that we plan to clean up the design of the commands later. I have no idea or no comment on whether `write_mutations` will be reused, I initially thought that we will rather be introducing new commands and deprecating the old ones, but it's probably an unimportant detail from the perspective of now.

Jadw1 replied:
> In that case I believe they won't affect current use because we only reload service levels cache when there are mutations that modify the service levels table. @Jadw1 please confirm.

That's right. I'm not going to move topology or schema changes to `write_mutations`.

Jadw1 replied:
To sum up the conversation, I'm following the solution with looking at table_ids in `write_mutations` and I'm not adding any new command type for service levels.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9658/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9658/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9658/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2046/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10069/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9658/)
#### Build Details:
- Duration: 2 hr 43 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9666/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9666/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9666/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2052/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10075/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9666/)
#### Build Details:
- Duration: 2 hr 44 min
- Builder: spider5.cloudius-systems.com

kostja commented on service/qos/service_level_controller.cc:
```diff
@@ -221,6 +223,25 @@ future<> service_level_controller::update_service_levels_from_distributed_data()
+    assert(this_shard_id() == global_controller);
+
+    if (storage_service.get_topology_upgrade_state() != service::topology::upgrade_state_type::done || group0_client.in_recovery()) {
+        update_from_distributed_data([cfg] () {
```
I undestand update_from_distributed_data() is from the old design, but the code now is not obvoius. maybe_start() calls update(), and doesn't start anything. Ideally it should start the loop as its name prescribes. Can update_from_distributed_data() be inlined into maybe_start()?

Jadw1 replied:
I've merged those 2 functions into one and renamed `update_from_distributed_data()` to `maybe_start_legacy_update_from_distributed_data()`, so the name better reflects what the method does.

kostja commented on service/qos/service_level_controller.cc:
```diff
@@ -307,18 +328,25 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
                     f.get();
+                    
+                    if (storage_service.get_topology_upgrade_state() == service::topology::upgrade_state_type::done) {
+                        stop_update_from_distributed_data();
```
Shouldn't it be claled stop_legacy_update_loop() for symmetry?

gleb-cloudius commented on service/raft/group0_state_machine.hh:
```diff
@@ -94,6 +77,12 @@ struct group0_command {
+        service_levels_cache
+    };
+    using modules_to_reload = enum_set<super_enum<reload_module,
+        reload_module::service_levels_cache>>;
```
Why use `enum_set` here? Isn't the set known only in runtime?

piodul replied:
`enum_set` is not a compile-time set, it's just a dynamic set of bits based on the passed enum. It requires passing not a regular C++ enum but rather a `super_enum` which - for some reason - requires you to specify all possible variants you want to use in it as template parameters.

gleb-cloudius replied:
The comment in the code says:

```
/**
 *
 * Allows to take full advantage of compile-time information when operating
 * on a set of enum values.
 *      
 * Examples:
 *      
 *   enum class x { A, B, C };
 *   using my_enum = super_enum<x, x::A, x::B, x::C>;
 *   using my_enumset = enum_set<my_enum>; 
 *  
 *   static_assert(my_enumset::frozen<x::A, x::B>::contains<x::A>(), "it should...");
 *
 *   assert(my_enumset::frozen<x::A, x::B>::contains(my_enumset::prepare<x::A>()));
 *
 *   assert(my_enumset::frozen<x::A, x::B>::contains(x::A));
 *
 */

```

We do not need any of that and I find its usage to be just a code obfuscation to be honest.

piodul replied:
I suggested using it because previously we were using `std::unordered_set`, which seems quite heavy for the purpose of just keeping one or two integers. I don't think we have a less clunky abstraction which could let us implement a type-safe bitset.

Maybe we could turn this into a struct of bools?

```c++
struct modules_to_reload {
    bool service_levels_cache = false;
};
```

gleb-cloudius replied:
Yes, `std::unordered_set` is overkill. The struct is straightforward IMO.

avikivity replied:
It's a layering violation, no? It make group0 know about all its clients, but all its clients have to known about group0.

Better to register callbacks table_id -> callback. I don't think an unordered_set is a problem. The goal here is to disentangle the code, not save a few cycles in an administrative function.

Jadw1 replied:
Changed to use struct of bools instead of enum_set

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -667,6 +667,7 @@ future<> storage_service::topology_state_load() {
     co_await _sl_controller.invoke_on_all([this] (qos::service_level_controller& sl_controller) {
         sl_controller.upgrade_to_v2(_qp, _group0->client());
     });
+    co_await update_service_levels_cache();
```
I do not understand the commit message. It says "it needs to be
initialized on group0 start". This, here, is not group0 start.

Jadw1 replied:
You're right, fixed the commit message.

avikivity commented on service/raft/group0_state_machine.hh:
```diff
@@ -63,24 +64,6 @@ struct group0_command {
-    // they are trying to perform is no longer valid in the context of this new state).
-    //
-    // Otherwise we update the state ID (`new_state_id`).
-    //
```
Why is this comment removed?

Jadw1 replied:
idk, my mistake. restored

kostja commented:
* Avi Kivity ***@***.***> [24/06/20 17:13]:
> 
> > @@ -94,6 +77,12 @@ struct group0_command {
>  // Raft state machine implementation for managing group 0 changes (e.g. schema changes).
>  // NOTE: group 0 raft server is always instantiated on shard 0.
>  class group0_state_machine : public raft_state_machine {
> +    enum class reload_module {
> +        service_levels_cache
> +    };
> +    using modules_to_reload = enum_set<super_enum<reload_module,
> +        reload_module::service_levels_cache>>;
> 
> It's a layering violation, no? It make group0 know about all its clients, but all its clients have to known about group0.
> 
> Better to register callbacks table_id -> callback. I don't think
> an unordered_set is a problem. The goal here is to disentangle
> the code, not save a few cycles in an administrative function.

What sort of table id do you have in mind? The idea is to process
all changes for all tables at once, not one table at a time.

We can apply multiple entires from the raft log in a single go.


-- 
Konstantin Osipov, Moscow, Russia

kostja commented:
* Avi Kivity ***@***.***> [24/06/20 17:13]:
> > @@ -94,6 +77,12 @@ struct group0_command {
>  // Raft state machine implementation for managing group 0 changes (e.g. schema changes).
>  // NOTE: group 0 raft server is always instantiated on shard 0.
>  class group0_state_machine : public raft_state_machine {
> +    enum class reload_module {
> +        service_levels_cache
> +    };
> +    using modules_to_reload = enum_set<super_enum<reload_module,
> +        reload_module::service_levels_cache>>;
> 
> Although, I see the code isn't written this way. Maybe we should restructure it to react to table writes.

The idea is that we take a bunch of entries from the raft log,
build the new in-memory state based on them, and swap the current
state with the new one, not reload the in-memory state for each
entry.

-- 
Konstantin Osipov, Moscow, Russia

avikivity commented:
> * Avi Kivity ***@***.***> [24/06/20 17:13]:
> > @@ -94,6 +77,12 @@ struct group0_command { // Raft state machine implementation for managing group 0 changes (e.g. schema changes). // NOTE: group 0 raft server is always instantiated on shard 0. class group0_state_machine : public raft_state_machine { + enum class reload_module { + service_levels_cache + }; + using modules_to_reload = enum_set<super_enum<reload_module, + reload_module::service_levels_cache>>; It's a layering violation, no? It make group0 know about all its clients, but all its clients have to known about group0. Better to register callbacks table_id -> callback. I don't think an unordered_set is a problem. The goal here is to disentangle the code, not save a few cycles in an administrative function.
> What sort of table id do you have in mind? 

The table name.

> The idea is to process all changes for all tables at once, not one table at a time. We can apply multiple entires from the raft log in a single go.

A module can register interest in one or more tables. If the log contains changes to any of those tables, we tell it to reload. If the log doesn't contain changes to any of the tables it's interested in, we don't tell it to reload.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10001/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10001/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10001/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2298/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10316/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10001/)
#### Failed Tests (57/32562):
- [test_short_read_delete](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read_delete) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_delete)
- [test_data_query_diges](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_data_query_diges) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_query_diges)
- [test_incomplete_result_test_partition_limit](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_incomplete_result_test_partition_limit) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incomplete_result_test_partition_limit)
- [test_incomplete_result_test_per_partition_row_limit](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_incomplete_result_test_per_partition_row_limit) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incomplete_result_test_per_partition_row_limit)
- [test_reaching_end_after_retry](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_reaching_end_after_retry) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_reaching_end_after_retry)
- [test_short_read](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read)
- [test_short_read_reversed](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read_reversed) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_reversed)
- [test_quorum_with_null](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/cql_additional_tests/TestCQL/Tests___dtest_with_topology_changes___test_quorum_with_null) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_quorum_with_null)
- [test_cluster_patient_cql_connection_after_node_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/cql_session_test/TestCqlSession/Tests___dtest_with_topology_changes___test_cluster_patient_cql_connection_after_node_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_patient_cql_connection_after_node_restart)
- [test_truncate_before_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10001/testReport/junit/cql_tests/TestTruncate/Tests___dtest_with_topology_changes___test_truncate_before_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_before_restart)
#### Build Details:
- Duration: 2 hr 48 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10034/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10034/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10034/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2314/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10333/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10034/)
#### Failed Tests (40/32508):
- [test_short_read_delete](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_short_read_delete) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_short_read_delete)
- [test_quorum_available_during_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_quorum_available_during_failure) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_quorum_available_during_failure)
- [test_data_query_diges](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_data_query_diges) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_query_diges)
- [test_incomplete_result_test_partition_limit](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_incomplete_result_test_partition_limit) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incomplete_result_test_partition_limit)
- [test_incomplete_result_test_per_partition_row_limit](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_incomplete_result_test_per_partition_row_limit) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incomplete_result_test_per_partition_row_limit)
- [test_reaching_end_after_retry](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/consistency_test/TestConsistency/Tests___dtest_with_topology_changes___test_reaching_end_after_retry) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_reaching_end_after_retry)
- [test_cluster_patient_cql_connection_after_node_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/cql_session_test/TestCqlSession/Tests___dtest_with_topology_changes___test_cluster_patient_cql_connection_after_node_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_patient_cql_connection_after_node_restart)
- [test_after_stop_start_value_is_default](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/nodetool_additional_test/TestGetTraceProbability/Tests___dtest_with_topology_changes___test_after_stop_start_value_is_default) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_after_stop_start_value_is_default)
- [test_repair_cell_delete](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_cell_delete) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_cell_delete)
- [test_repair_partition_delete](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10034/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_partition_delete) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_partition_delete)
#### Build Details:
- Duration: 2 hr 35 min
- Builder: spider6.cloudius-systems.com

Jadw1 commented:
Changelog:
- fixed bug which caused the failing tests
- updated `write_mutations` description
- changed `modules_to_update` in `group0_state_machine` to be struct of bools
- addressed rest of comments
- improved comments in the code and commits messages

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10063/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10063/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10063/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2341/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10363/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10063/)
#### Build Details:
- Duration: 3 hr 42 min
- Builder: spider5.cloudius-systems.com

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -221,6 +221,20 @@ future<> service_level_controller::update_service_levels_from_distributed_data()
             }
         });
+    }).handle_exception([this] (std::exception_ptr eptr) {
+        try {
+            std::rethrow_exception(eptr);
+        } catch (const broken_semaphore& ex) {
+            // Catch broken semaphore exception.
+            // If it was caused by early abort subscription, ignore the exception.
+            // Otherwise, rethrow it.
+
+            if (!*_early_abort_subscription) {
+                // Early abort subscription was fired
+                return make_ready_future();
+            }
+            std::rethrow_exception(std::move(eptr));
+        }
```
I'm not a fan of this solution. It relies on non-local reasoning (it assumes that if there is a `broken_semaphore` exception and `!*_early_abort_subscription`, then it must be correlated).

It looks like we are using the `notifications_serializer` semaphore for two reasons:

1. To serialize operations on service levels,
2. To prevent more operations from being done on service levels when we are shutting down.

It currently fits purpose (1) well and we do want to serialize the current operations as we currently do (i.e. reloading SL state, adding/removing SLs when an incoming SL is not known). However, I think there is a problem with purpose (2). I think it would be much cleaner to add a `gate` and make it responsible for (2). You could remove the `.broken()` call in `do_abort()`, and only close the gate in `stop()`.

avikivity replied:
A solution I like (don't know it it applies well here) is to use an update loop that waits on a condition variable and an abort_source. When the condition variable is signaled, it reads and applies the changes. The service waits to the update loop in its stop() method.

avikivity replied:
This has the nice property that if the update loop is slow, then multiple updates are coalesced.

piodul replied:
The current update loop is being deprecated so we are not going to change the algorithm. Instead, the SL state update will be triggered by group0. I agree that it's a nice way to write an update loop, though.

The problem with the code in the PR right now is that the semaphore is broken when the global abort_source is being triggered. This semaphore is taken in the function that reloads the SL state and which is invoked by group0 code, so the exception from the broken semaphore breaks the group0 state machine.

I suggested splitting the semaphore into a semaphore + gate, but perhaps we could just "break" the semaphore in service_level_controller::stop()...? (AFAIK it is called after group0 is stopped, so it should work). I'm not familiar with using broken semaphores to control access to a component, though, so I suggested using a gate.

avikivity replied:
I wasn't referring to the current update loop. Just a nice pattern to serialize and batch events.

Perhaps we can capture it in some template so it can be easily reused.

piodul reviewed (changes requested):
Looks good, but I don't like how the broken_semaphore exception is handled. I suggest changing how the notification serializer semaphore is used - see my inline comments.

Jadw1 commented on service/qos/service_level_controller.cc:
```diff
@@ -125,7 +123,8 @@ future<> service_level_controller::stop() {
     }
-
+    
+    _global_controller_db->notifications_serializer.broken();
```
This is followup to https://github.com/scylladb/scylladb/pull/18758#discussion_r1669802920 conversation, since I've dropped that commit.

I don't think the semaphore is used to (2) control an access to service levels. 
To me it looks like it's only used to serialize write access to sl cache (as the name suggests). Nowhere in the code methods which uses the semaphore are in try/catch blocks and I also didn't find any mentions that the semaphore control the access to service levels.

So I've moved semaphore `broken()` to the `stop()` method, which should fix the issue with crash in `topology_state_load()`.

piodul replied:
Ok, so why do we break this semaphore at all? Why was this call introduced in the first place? Maybe we can just remove `.broken()` altogether?

I'd rather not remove it without understanding why it was introduced in the first place because of the [Chesterton's fence](https://en.wikipedia.org/wiki/G._K._Chesterton#Chesterton's_fence) principle.

We can handle this in a followup, though.

Jadw1 commented:
Changelog:
- removed commit with catching broken semaphore exception and moved semaphore breaking to stop()

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10310/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10310/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10310/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2535/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10559/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10310/)
#### Failed Tests (1/32675):
- [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10310/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_3)
#### Build Details:
- Duration: 2 hr 42 min
- Builder: spider3.cloudius-systems.com

piodul commented:
> #### Failed Tests (1/32675):
> * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10310/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_3)
> 

@Jadw1 please take a look at the failure. Is this a known issue?

mykaul commented:
> > #### Failed Tests (1/32675):
> > 
> > * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10310/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_3)
> 
> @Jadw1 please take a look at the failure. Is this a known issue?

https://github.com/scylladb/scylladb/issues/19244 ?

Jadw1 commented:
> > > #### Failed Tests (1/32675):
> > > 
> > > * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10310/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_3)
> > 
> > 
> > @Jadw1 please take a look at the failure. Is this a known issue?
> 
> #19244 ?

Yeah, it looks like. Retriggering CI

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10352/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10352/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_service_levels
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10352/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2560/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10581/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10352/)
#### Build Details:
- Duration: 2 hr 39 min
- Builder: spider3.cloudius-systems.com

piodul commented:
Queued
